### DISCUSSION TITLE
How are the functions implemented in io.h?

### DISCUSSION THREAD

**Discussion by DenisVASI9** [ipxe/ipxe#953] [Q&A]
Hi, I'm debugging ipxe and I don't understand how the functions in io.h file are implemented, for example:
void writel ( uint32_t data, volatile uint32_t *io_addr );
#define writel( data, io_addr ) \
  IOAPI_WRITE ( writel, uint32_t, data, io_addr, "MEM", 8 )

I only see the header writel, but I don't understand where the implementation is. 
Could you please tell me

**Comment by mcb30** [comment]
For x86, the implementation is in `arch/x86/include/ipxe/x86_io.h`.  `writel()` is constructed via a macro:
```c
X86_WRITEX ( writel, uint32_t );
```
